backend/lucene/src/main/java/org/hibernate/search/backend/lucene/document/impl/AbstractLuceneNonFlattenedDocumentBuilder.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.document.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.search.backend.lucene.document.model.impl.LuceneIndexSchemaObjectNode;
+import org.hibernate.search.backend.lucene.logging.impl.Log;
+import org.hibernate.search.backend.lucene.multitenancy.impl.MultiTenancyStrategy;
+import org.hibernate.search.backend.lucene.util.impl.LuceneFields;
+import org.hibernate.search.util.common.logging.impl.LoggerFactory;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.index.IndexableField;
+
+/**
+ * @author Guillaume Smet
+ */
+abstract class AbstractLuceneNonFlattenedDocumentBuilder extends AbstractLuceneDocumentBuilder
+		implements LuceneDocumentBuilder {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	final Document document = new Document();
+	private final Map<String, EncounteredFieldStatus> fieldStatus = new HashMap<>();
+
+	AbstractLuceneNonFlattenedDocumentBuilder(LuceneIndexSchemaObjectNode schemaNode) {
+		super( schemaNode );
+	}
+
+	@Override
+	public void addField(IndexableField field) {
+		document.add( field );
+	}
+
+	@Override
+	public void addFieldName(String absoluteFieldPath) {
+		// If the status was already ENCOUNTERED, just replace it.
+		fieldStatus.put( absoluteFieldPath, EncounteredFieldStatus.ENCOUNTERED_AND_NAME_INDEXED );
+	}
+
+	@Override
+	void checkNoValueYetForSingleValued(String absoluteFieldPath) {
+		EncounteredFieldStatus previousValue = fieldStatus.putIfAbsent( absoluteFieldPath, EncounteredFieldStatus.ENCOUNTERED );
```

Comment:
Is it possible that we have multiple concurrent requests here?

Reply:
Nope, the document builder is single-threaded. I just need `putIfAbsent` because of its semantics, not because I need atomicity.